packages/eslint-plugin/src/rules/no-floating-promises.ts

```diff
@@ -327,10 +329,8 @@ export default createRule<Options, MessageId>({
         return isUnhandledPromise(checker, node.right);
       }
 
-      // We conservatively return false for all other types of expressions because
-      // we don't want to accidentally fail if the promise is handled internally but
-      // we just can't tell.
-      return { isUnhandled: false };
+      // Anything else is unhandled.
+      return { isUnhandled: true };
```

Comment:
In your summary you wrote

> Changes implementation from an unhandled promise syntax allowlist to a syntax denylist.

But I don't understand what we gain from that switch

This seems... dangerous - switching to default assuming it's unhandled is probably going to lead to false positives at scale.

Reply:
> I don't understand what we gain from that switch
 
implementation simplicity and future proofing 🤷  It seems more straightforward, and more correct (IMO), to opt _out_ of flagging behaviors indicated by the docs than it does to opt _in_ to the ones we do want to flag...

Nodes that are handled
- `await` expression
- assignment expression (x = y)

Nodes that require special handling to determine whether they're handled or unhandled (and therefore need to be specified individually no matter what)
- CallExpression
- LogicalExpression
- SequenceExpression
- ConditionalExpression
- `void` expression
- optional chaining expression
- iifes

Nodes that are always unhandled (if they result in a promise)
- NewExpression
- Identifier
- MemberExpression
- TaggedTemplateExpression (previously forgotten, recently fixed in https://github.com/typescript-eslint/typescript-eslint/pull/8758)
- TSAsExpression (previously forgotten, the point of this PR)
- TSTypeAssertion (previously forgotten, the point of this PR)
- TSNonNullExpression (previously forgotten, basically equivalent to previous two)
- ... probably more - this feels like the appropriate default behavior for a rule that doesn't fall into one of the other two buckets

> This seems... dangerous - switching to default assuming it's unhandled is probably going to lead to false positives at scale.

To be sure, I gave this some thought before making this change, and I just couldn't think of any scenarios where this would really be problematic... Every case I thought of seeed to be beneficial. Just to be sure we're on the same page, that code is only reached _if the expression results in a promise_. 

That does mean that this code will start flagging now
```ts
({
  then(resolve: Function, reject: Function) {
    return this;
  }
});
```
and I think that makes sense, given 
```ts
const unhandled = {
  then(resolve: Function, reject: Function) {
    return this;
  }
};
unhandled;
```
already flags (of course, this touches upon https://github.com/typescript-eslint/typescript-eslint/issues/8433). But it's not like we're suddenly going to flag every object literal; just thenable ones.

I think the `yield` cases added to the tests make sense to flag as well....

Note that we dogfood `no-floating-promises` in this repo and this change didn't result in _any_ new reports. 

My conclusion - false positives are a very valid concern to explore, but for now I do think that this change is safe/beneficial. Would definitely be open to being proven wrong, though, especially if a good counterexample comes to mind! 🙂